AI/BattleAI/BattleAI.cpp

```diff
@@ -280,7 +325,47 @@ BattleAction CBattleAI::goTowards(const CStack * stack, const battle::Unit * ene
 
 BattleAction CBattleAI::useCatapult(const CStack * stack)
 {
-	throw std::runtime_error("CBattleAI::useCatapult is not implemented.");
+	BattleAction attack;
+	BattleHex targetHex = BattleHex::INVALID;
+
+	if(cb->battleGetGateState() == EGateState::CLOSED)
+	{
+		targetHex = cb->wallPartToBattleHex(EWallPart::GATE);
+	}
+	else
+	{
+		EWallPart::EWallPart wallParts[] = {
+			EWallPart::KEEP,
+			EWallPart::BOTTOM_TOWER,
+			EWallPart::UPPER_TOWER,
+			EWallPart::BELOW_GATE,
+			EWallPart::OVER_GATE,
+			EWallPart::BOTTOM_WALL,
+			EWallPart::UPPER_WALL
+		};
+
+		for(auto wallPart : wallParts)
+		{
+			auto wallState = cb->battleGetWallState(wallPart);
+
+			if(wallState != EWallState::DESTROYED)
+			{
+				targetHex = cb->wallPartToBattleHex(wallPart);

```

Comment:
Should be `break` here

Reply:
Invert if to reduce nesting?

Reply:
Do not understand how break can be used here. I need all broken wall hexes because goTowardsNearest will select the nearest one

Reply:
You are cycling all wall parts and select last not destroyed, if you really need last one, set `wallParts` in reverse order and break `wallPart : wallParts` cycle after first match.

Reply:
Not last but all destroyed parts so I can pass through the one which is nearest.

Reply:
This is balista. My fault. I thought it was another method.

Reply:
I do not plan to implement something very clever and arguing what has to be broken first walls or towers. My idea was simple. Break gate and afterwards break towers so that they do not give extra damage. Btw need to test ballista for fort which has no towers

Reply:
If you have better idea in which order it is better to break walls in any situation I can accept it so. The main idea of all these changes is to reduce probability of freeze because of endless battle.

Reply:
Break added